modules/service/src/main/scala/lucuma/odb/service/ProposalAttachmentFileService.scala

```diff
@@ -218,51 +220,43 @@ object ProposalAttachmentFileService extends AttachmentFileService {
         user: User,
         programId: Program.Id,
         attachmentType: Tag
-      ): F[Either[AttachmentException, Stream[F, Byte]]] = 
-        session.transaction
-          .use(_ =>
-            for {
-              _    <- checkAccess(session, user, programId)
-              _    <- checkAttachmentType(attachmentType)
-              path <- getAttachmentRemotePathFromDB(user, programId, attachmentType)
-            } yield path
-          )
-          .flatMap { rpath =>
-            s3FileSvc.verifyAndGet(rpath).map(_.asRight)
-          }
-          .recover {
-            case e: AttachmentException => e.asLeft
-          }
+      )(using NoTransaction[F]): F[Either[AttachmentException, Stream[F, Byte]]] = {
+          for {
+            path <- services.transactionally {
+              checkAccess(session, user, programId) >>
+              checkAttachmentType(attachmentType) >>
+              getAttachmentRemotePathFromDB(user, programId, attachmentType)
+            }
+            res <- s3FileSvc.verifyAndGet(path).map(_.asRight)
+        } yield res
+      } recover {
+        case e: AttachmentException => e.asLeft
+      }
 
       def insertAttachment(
         user: User,
         programId: Program.Id,
         attachmentType: Tag,
         fileName: String,
         description: Option[NonEmptyString],
-        data: Stream[F, Byte]): F[Unit] = 
+        data: Stream[F, Byte])(using NoTransaction[F]): F[Unit] = 
         FileName
           .fromString(fileName)
           .fold(
-            e  => Async[F].raiseError(e),
+            e  => MonadCancelThrow[F].raiseError(e),
             fn =>
-              session.transaction
-                .use(_ =>
-                  for {
-                    _ <- checkAccess(session, user, programId)
-                    _ <- checkAttachmentType(attachmentType)
-                    _ <- checkForDuplicateAttachment(user, programId, attachmentType)
-                    _ <- checkForDuplicateName(programId, fn, none)
-                  } yield ()
-                )
-                .flatMap( _ =>
-                  for {
-                    uuid   <- Async[F].delay(UUID.randomUUID())
-                    path    = filePath(programId, uuid, fn.value)
-                    size   <- s3FileSvc.upload(path, data)
-                    result <- insertAttachmentInDB(user, programId, attachmentType, fn, description, size, path)
-                  } yield result
-                )
+              for {
+                _      <- services.transactionally {
```

Comment:
I suppose this one is different (`using NoTransaction[F]` and then `services.transactionally`) to not hold a transaction open while it does a file upload?  What would happen in the theoretical case of another service method that is `using Transaction[F]` calling `insertAttachment`?  I guess it doesn't compile?

Reply:
Correct. If your transaction status is part of the type it will prevent you from composing things the wrong way. It's a weak guarantee because nothing is forcing you to put those things on the types, but it seems like a reasonable pattern.